File: back/domain/model/external_service_url.go
package model

import (
	"errors"
	"fmt"
	"net/url"
)

const (
	ExternalServiceOther = iota
	ExternalServiceGithub
	ExternalServiceX
	ExternalServiceQiita
	ExternalServiceZenn
	ExternalServiceNote
)

type ExternalServiceUrl struct {
	url         string
	serviceType int
	id          string
}

func NewExternalServiceUrl(id string, serviceType int, url string) (*ExternalServiceUrl, error) {
	if len(url) == 0 {
		return nil, errors.New("URLを入力してください")
	}

	updatedServiceType, err := updateServiceType(serviceType)

	if err != nil {
		return nil, err
	}

	updatedUrl, err := updateUrlLogic(url, updatedServiceType)

	if err != nil {
		return nil, err
	}

	return &ExternalServiceUrl{serviceType: updatedServiceType, url: updatedUrl, id: id}, nil
}

func updateUrlLogic(urlRaw string, serviceType int) (string, error) {
	_, err := updateServiceType(serviceType) // serviceTypeのvalidate

	if err != nil {
		return "", err
	}

	// URLの形式が異なっている場合エラー
	parsedURL, err := url.ParseRequestURI(urlRaw)
	if err != nil {
		return "", fmt.Errorf("無効なURL形式です: %w", err)
	}
	// スキーム（http/httpsなど）とホストが存在しているかもチェックする
	if parsedURL.Scheme == "" || parsedURL.Host == "" {
		return "", fmt.Errorf("URLにスキームまたはホストが不足しています")
	}

	hostname := parsedURL.Hostname()

	type combination struct {
		serviceType int
		hostname    string
	}

	combinations := []combination{
		{
			serviceType: ExternalServiceGithub,
			hostname:    "github.com",
		},
		{
			serviceType: ExternalServiceX,
			hostname:    "x.com",
		},
		{
			serviceType: ExternalServiceZenn,
			hostname:    "zenn.dev",
		},
		{
			serviceType: ExternalServiceNote,
			hostname:    "note.com",
		},
		{
			serviceType: ExternalServiceQiita,
			hostname:    "qiita.com",
		},
	}

	for _, c := range combinations {
		if c.serviceType == serviceType && c.hostname != hostname {
			return "", fmt.Errorf("servicetypeは%dですが、ホスト名が%sで異なります", serviceType, hostname)
		}
	}

	return urlRaw, nil
}

func updateServiceType(serviceType int) (int, error) {
	switch serviceType {
	case ExternalServiceOther:
	case ExternalServiceGithub:
	case ExternalServiceX:
	case ExternalServiceQiita:
	case ExternalServiceZenn:
	case ExternalServiceNote:
	default:
		return 0, fmt.Errorf("unknown service type:%d", serviceType)
	}
	return serviceType, nil
}

func (e *ExternalServiceUrl) UpdateUrl(url string) error {
	updatedUrl, err := updateUrlLogic(url, e.serviceType)

	if err != nil {
		return err
	}

	e.url = updatedUrl

	return nil
}

func (e *ExternalServiceUrl) UpdateServiceType(serviceType int) error {
	updatedServiceType, err := updateServiceType(serviceType)
	if err != nil {
		return err
	}

	e.serviceType = updatedServiceType
	return nil
}

func (e *ExternalServiceUrl) Url() string {
	return e.url
}

func (e *ExternalServiceUrl) ServiceType() int {
	return e.serviceType
}

func (e *ExternalServiceUrl) ID() string {
	return e.id
}

File: front/lib/externalServiceUrl.ts
/**
const (
	ExternalServiceOther = iota
	ExternalServiceGithub
	ExternalServiceX
	ExternalServiceQiita
	ExternalServiceZenn
	ExternalServiceNote
)
 */

export const MAX_SERVICE_TYPE_LEN = 5

export function ServiceTypeToString(serviceType: number) {
	switch (serviceType) {
		case 0:
			return 'other'
		case 1:
			return 'github'
		case 2:
			return 'x'
		case 3:
			return 'qiita'
		case 4:
			return 'zenn'
		case 5:
			return 'note'
		default:
			return 'other'
	}
}

export function ServiceTypeToServiceName(serviceType: number) {
	switch (serviceType) {
		case 1:
			return 'GitHub'
		case 2:
			return 'X'
		case 3:
			return 'Qiita'
		case 4:
			return 'Zenn'
		case 5:
			return 'note'
		default:
			return 'other'
	}
}

export function StringToServiceType(serviceType: string) {
	switch (serviceType) {
		case 'github':
			return 1
		case 'x':
			return 2
		case 'qiita':
			return 3
		case 'zenn':
			return 4
		case 'note':
			return 5
		default:
			return 0
	}
}

export function ImageNameByServiceType(serviceType: number) {
	switch (serviceType) {
		case 1:
			return 'github.png'
		case 2:
			return 'x.png'
		case 3:
			return 'qiita.png'
		case 4:
			return 'zenn.svg'
		case 5:
			return 'note.png'
		default:
			return 'other.svg'
	}
}

export function ConvertToExternalServiceUrl(serviceType: number, name: string): string {
	if (name === '') return ''
	switch (serviceType) {
		case 1:
			return `https://github.com/${name}`
		case 2:
			return `https://x.com/${name}`
		case 3:
			return `https://qiita.com/${name}`
		case 4:
			return `https://zenn.dev/${name}`
		case 5:
			return `https://note.com/${name}`
		default:
			return ''
	}
}

export function ConvertUrlToServiceUserName(url: string): string {
	try {
		const parsedUrl = new URL(url);
		const host = parsedUrl.host;
		if (host === 'github.com') {
			return parsedUrl.pathname.split('/')[1];
		} else if (host === 'x.com') {
			return parsedUrl.pathname.split('/')[1];
		} else if (host === 'qiita.com') {
			return parsedUrl.pathname.split('/')[1];
		} else if (host === 'zenn.dev') {
			return parsedUrl.pathname.split('/')[1];
		} else if (host === 'note.com') {
			return parsedUrl.pathname.split('/')[1];
		}
	} catch (e) {
		// Handle invalid URL
		console.error('Invalid URL:', e);
	}
	return url;
}
